src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUnitTest.java

```diff
@@ -1,91 +1,1138 @@
+
+        /* update x and y -- f, g and h must go out of date */
+        ImmutableComputableGraph icg_tmp_xy = icg_tmp
+                .setValue("x", new DuplicableNDArray(getRandomINDArray()))
+                .setValue("y", new DuplicableNumber<>(getRandomDouble()));
+        Assert.assertTrue(!icg_tmp_xy.isValueDirectlyAvailable("f"));
+        Assert.assertTrue(!icg_tmp_xy.isValueDirectlyAvailable("g"));
+        Assert.assertTrue(!icg_tmp_xy.isValueDirectlyAvailable("h"));
+
+        /* update x and z -- f, g and h must go out of date */
+        ImmutableComputableGraph icg_tmp_xz = icg_tmp
+                .setValue("x", new DuplicableNDArray(getRandomINDArray()))
+                .setValue("z", new DuplicableNDArray(getRandomINDArray()));
+        Assert.assertTrue(!icg_tmp_xz.isValueDirectlyAvailable("f"));
+        Assert.assertTrue(!icg_tmp_xz.isValueDirectlyAvailable("g"));
+        Assert.assertTrue(!icg_tmp_xz.isValueDirectlyAvailable("h"));
+
+        /* update x and z -- f, g and h must go out of date */
+        ImmutableComputableGraph icg_tmp_xyz = icg_tmp
+                .setValue("x", new DuplicableNDArray(getRandomINDArray()))
+                .setValue("y", new DuplicableNumber<>(getRandomDouble()))
+                .setValue("z", new DuplicableNDArray(getRandomINDArray()));
+        Assert.assertTrue(!icg_tmp_xyz.isValueDirectlyAvailable("f"));
+        Assert.assertTrue(!icg_tmp_xyz.isValueDirectlyAvailable("g"));
+        Assert.assertTrue(!icg_tmp_xyz.isValueDirectlyAvailable("h"));
+
+        if (f_external) {
+            /* update f -- h must go out of date */
+            ImmutableComputableGraph icg_tmp_f = icg_tmp
+                    .setValue("f", new DuplicableNDArray(getRandomINDArray()));
+            Assert.assertTrue(!g_caching || icg_tmp_f.isValueDirectlyAvailable("g"));
+            Assert.assertTrue(!icg_tmp_f.isValueDirectlyAvailable("h"));
+        }
+
+        if (g_external) {
+            /* update g -- h must go out of date */
+            ImmutableComputableGraph icg_tmp_g = icg_tmp
+                    .setValue("g", new DuplicableNDArray(getRandomINDArray()));
+            Assert.assertTrue(!f_caching || icg_tmp_g.isValueDirectlyAvailable("f"));
+            Assert.assertTrue(!icg_tmp_g.isValueDirectlyAvailable("h"));
+        }
+
+        if (f_external && g_external) {
+            /* update f and g -- h must go out of date */
+            ImmutableComputableGraph icg_tmp_fg = icg_tmp
+                    .setValue("f", new DuplicableNDArray(getRandomINDArray()))
+                    .setValue("g", new DuplicableNDArray(getRandomINDArray()));
+            Assert.assertTrue(!icg_tmp_fg.isValueDirectlyAvailable("h"));
+        }
+    }
+
+    /**
+     * Tests propagation of tags from descendents to parents
+     */
+    @Test(dataProvider = "allPossibleNodeFlags", invocationCount = NUM_TRIALS)
+    public void testTagPropagation(final boolean f_caching, final boolean f_external,
```

Comment:
This test is actually redundant given the tests in `ComputableGraphStructureUnitTest` but I thought of keeping it because it is a _concrete_ test for the _concrete_ ICG worked out in this unit test. It may help some person in the future to understand what tag propagation means, etc.